controllers/statusmanager/status_manager.go

```diff
@@ -5,64 +5,315 @@ package statusmanager
 
 import (
 	"context"
-	"time"
+	"fmt"
+	"reflect"
+	"strings"
+	"sync"
 
 	configv1 "github.com/openshift/api/config/v1"
-	"github.com/openshift/cluster-network-operator/pkg/controller/statusmanager"
+	"github.com/openshift/library-go/pkg/config/clusteroperator/v1helpers"
+	"gopkg.in/yaml.v2"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/meta"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/util/retry"
 	"sigs.k8s.io/controller-runtime/pkg/client"
 	logf "sigs.k8s.io/controller-runtime/pkg/log"
 
 	operatorv1 "github.com/vmware/antrea-operator-for-kubernetes/api/v1"
+	"github.com/vmware/antrea-operator-for-kubernetes/controllers/sharedinfo"
 	operatortypes "github.com/vmware/antrea-operator-for-kubernetes/controllers/types"
+	"github.com/vmware/antrea-operator-for-kubernetes/version"
 )
 
 var log = logf.Log.WithName("status_manager")
 
-func SetAntreaInstallStatus(cl client.Client, conditionType configv1.ClusterStatusConditionType, status configv1.ConditionStatus, t time.Time, reason, message string) {
+type StatusLevel int
+
+const (
+	ClusterConfig StatusLevel = iota
+	OperatorConfig
+	PodDeployment
+	RolloutHung
+	ClusterNode
+	maxStatusLevel
+)
+
+type Adaptor interface {
+	getLastPodState(status *StatusManager) (map[types.NamespacedName]daemonsetState, map[types.NamespacedName]deploymentState)
+	setLastPodState(status *StatusManager, dss map[types.NamespacedName]daemonsetState, deps map[types.NamespacedName]deploymentState) error
+	set(status *StatusManager, reachedAvailableLevel bool, conditions ...configv1.ClusterOperatorStatusCondition)
+}
+
+// Status coordinates changes to AntreaInstall.status and ClusterOperator.Status.
+type StatusManager struct {
+	sync.Mutex
+
+	client client.Client
+	mapper meta.RESTMapper
+
+	name    string
+	version string
+
+	failing [maxStatusLevel]*configv1.ClusterOperatorStatusCondition
+
+	daemonSets  []types.NamespacedName
+	deployments []types.NamespacedName
+
+	OperatorNamespace string
+	AdaptorName       string
+	Adaptor
+}
+
+type StatusK8s struct{}
+
+type StatusOc struct{}
+
+func New(client client.Client, mapper meta.RESTMapper, name, operatorNamespace, version string, sharedInfo *sharedinfo.SharedInfo) *StatusManager {
+	status := StatusManager{
+		client:            client,
+		mapper:            mapper,
+		name:              name,
+		version:           version,
+		OperatorNamespace: operatorNamespace,
+		AdaptorName:       sharedInfo.AntreaPlatform,
+	}
+	if sharedInfo.AntreaPlatform == "openshift" {
```

Comment:
Do you think it might be worth having users explicitly specify either "openshift" or "k8s".
the reason is that if one mistypes "Openshift" "OpenShift" or other more or less funny misspellings of this word, the operator would load the K8s adaptor and the user wouldn't realize it.

Reply:
Agree with this, we better explicitly specify "openshift" or "k8s" to avoid misspellings and confusion.